File: swt/MFD/SRC/LCL/STD.R/INTEL.R
#     Copyright (c) 1979 by Allen Akin and Tom Chappell.

# These programs were developed by Allen Akin and Tom Chappell
# for  use  with  the  Software Tools Subsystem at the Georgia
# Institute of Technology.  General permission is  granted  to
# copy  and  use  this  software  product,  provided that this
# notice is included in full with each copy.



# intel --- generate Intel object tape from 8080 'qda' file

include PRIMOS_KEYS
include "as6800_def.r.i"

define(MEMSIZE,16384)
define(RECSIZE,16)

   integer fd, address, i, code, size
   integer open, getarg, gctoi

   character arg (MAXLINE)

   string default "l.out"

   call dsinit (MEMSIZE)

   if (getarg (1, arg, MAXLINE) == EOF)
      call scopy (default, 1, arg, 1)

   fd = open (arg, READ)
   if (fd == ERR)
      call cant (arg)

   if (getarg (2, arg, MAXLINE) == EOF)
      address = 0
   else {
      i = 1
      address = gctoi (arg, i, 16)
      }

   call load (fd, code, size)
   call relocate (fd, code, size, address)
   call dump (code, size, address)

   call close (fd)
   stop
   end



# load --- load a code segment from file fd, return its locn and size

   subroutine load (fd, code, size)
   integer fd, size
   pointer code

   include "intel_com.r.i"

   integer i, b
   integer getbyte

   pointer dsget

   if (getbyte (b, fd) == EOF || b ~= TEXTCODE)
      call error ("badly formed code file.")
   call getword (size, fd)

   code = dsget (size)
   for (i = 0; i < size; i = i + 1)
      call getbyte (mem (code + i), fd)

   return
   end



# relocate --- relocate code segment to specified address

   subroutine relocate (fd, code, size, address)
   integer fd, size, address
   pointer code

   include "intel_com.r.i"

   integer i, map, addr, unnecessary
   integer getbyte

   if (getbyte (i, fd) == EOF || i ~= RMAPCODE)
      call error ("badly formed code file.")
   call getword (i, fd)

   for (i = 0; i < size; i = i + 1) {
      if (mod (i, 8) == 0)
         call getbyte (map, fd)
      if (and (map, ls (1, 7 - mod (i, 8))) ~= 0) {
         addr = address + mem (code + i) + ls (mem (code + i + 1), 8)
         unnecessary = code + i
         mem (unnecessary) = rt (addr, 8)
         unnecessary = unnecessary + 1
         mem (unnecessary) = rs (addr, 8)
         }
      }

   return
   end



# dump --- write relocated program in hex to STDOUT

   subroutine dump (code, size, address)
   integer code, size, address

   include "intel_com.r.i"

   integer tw, mcont, cksum

   tw = 0

   while (tw < size) {
      call print (STDOUT, ":.")
      if (size - tw < RECSIZE)
         mcont = size - tw
      else
         mcont = RECSIZE
      cksum = 0
      call outhex (mcont, cksum)
      call outhex (rs (tw + address, 8), cksum)
      call outhex (rt (tw + address, 8), cksum)
      call outhex (0, cksum)
      repeat {
         call outhex (mem (tw + code), cksum)
         tw = tw + 1
         mcont = mcont - 1
         } until (mcont <= 0)
      call outhex (0 - cksum, cksum)
      call print (STDOUT, "*n.")
      }
   call print (STDOUT, ":00*n.")

   return
   end



# outhex --- print byte in hex on STDOUT, update checksum

   subroutine outhex (byte, checksum)
   integer byte, checksum

   string hex "0123456789ABCDEF"

   call putch (hex (rt (rs (byte, 4), 4) + 1), STDOUT)
   call putch (hex (rt (byte, 4) + 1), STDOUT)
   checksum = rt (checksum + byte, 8)

   return
   end



# getbyte --- read one byte from code file

   subroutine getbyte (b, fd)
   integer b, fd

   integer junk
   integer mapfd

   call prwf$$ (KREAD, mapfd (fd), loc (b), 1, intl (0), junk, junk)

   return
   end



# getword --- read word from object code file

   subroutine getword (w, fd)
   integer w, fd

   integer hi, lo

   call getbyte (lo, fd)
   call getbyte (hi, fd)
   w = or (ls (hi, 8), lo)

   return
   end

File: swt/src/lcl/std.r/intel.f
      INTEGER FD,ADDRE0,I,CODE,SIZE
      INTEGER OPEN,GETARG,GCTOI
      INTEGER ARG(102)
      INTEGER DEFAU0(6)
      DATA DEFAU0/236,174,239,245,244,0/
      CALL DSINIT(16384)
      IF((GETARG(1,ARG,102).NE.-1))GOTO 10000
        CALL SCOPY(DEFAU0,1,ARG,1)
10000 FD=OPEN(ARG,1)
      IF((FD.NE.-3))GOTO 10001
        CALL CANT(ARG)
10001 IF((GETARG(2,ARG,102).NE.-1))GOTO 10002
        ADDRE0=0
        GOTO 10003
10002   I=1
        ADDRE0=GCTOI(ARG,I,16)
10003 CALL LOAD(FD,CODE,SIZE)
      CALL RELOC0(FD,CODE,SIZE,ADDRE0)
      CALL DUMP(CODE,SIZE,ADDRE0)
      CALL CLOSE(FD)
      CALL SWT
      END
      SUBROUTINE LOAD(FD,CODE,SIZE)
      INTEGER FD,SIZE
      INTEGER CODE
      INTEGER MEM(16384)
      COMMON /DS$MEM/MEM
      INTEGER I,B
      INTEGER GETBY0
      INTEGER DSGET
      IF((GETBY0(B,FD).EQ.-1))GOTO 10005
      IF((B.NE.1))GOTO 10005
      GOTO 10004
10005   CALL ERROR('badly formed code file.')
10004 CALL GETWO0(SIZE,FD)
      CODE=DSGET(SIZE)
      I=0
      GOTO 10008
10006 I=I+1
10008 IF((I.GE.SIZE))GOTO 10007
        CALL GETBY0(MEM(CODE+I),FD)
      GOTO 10006
10007 RETURN
      END
      SUBROUTINE RELOC0(FD,CODE,SIZE,ADDRE0)
      INTEGER FD,SIZE,ADDRE0
      INTEGER CODE
      INTEGER MEM(16384)
      COMMON /DS$MEM/MEM
      INTEGER I,MAP,ADDR,UNNEC0
      INTEGER GETBY0
      IF((GETBY0(I,FD).EQ.-1))GOTO 10010
      IF((I.NE.2))GOTO 10010
      GOTO 10009
10010   CALL ERROR('badly formed code file.')
10009 CALL GETWO0(I,FD)
      I=0
      GOTO 10013
10011 I=I+1
10013 IF((I.GE.SIZE))GOTO 10012
        IF((MOD(I,8).NE.0))GOTO 10014
          CALL GETBY0(MAP,FD)
10014   IF((AND(MAP,LS(1,7-MOD(I,8))).EQ.0))GOTO 10015
          ADDR=ADDRE0+MEM(CODE+I)+LS(MEM(CODE+I+1),8)
          UNNEC0=CODE+I
          MEM(UNNEC0)=RT(ADDR,8)
          UNNEC0=UNNEC0+1
          MEM(UNNEC0)=RS(ADDR,8)
10015 GOTO 10011
10012 RETURN
      END
      SUBROUTINE DUMP(CODE,SIZE,ADDRE0)
      INTEGER CODE,SIZE,ADDRE0
      INTEGER MEM(16384)
      COMMON /DS$MEM/MEM
      INTEGER TW,MCONT,CKSUM
      TW=0
10016 IF((TW.GE.SIZE))GOTO 10017
        CALL PRINT(-11,':.')
        IF((SIZE-TW.GE.16))GOTO 10018
          MCONT=SIZE-TW
          GOTO 10019
10018     MCONT=16
10019   CKSUM=0
        CALL OUTHEX(MCONT,CKSUM)
        CALL OUTHEX(RS(TW+ADDRE0,8),CKSUM)
        CALL OUTHEX(RT(TW+ADDRE0,8),CKSUM)
        CALL OUTHEX(0,CKSUM)
10020     CALL OUTHEX(MEM(TW+CODE),CKSUM)
          TW=TW+1
          MCONT=MCONT-1
        IF((MCONT.GT.0))GOTO 10020
        CALL OUTHEX(0-CKSUM,CKSUM)
        CALL PRINT(-11,'*n.')
      GOTO 10016
10017 CALL PRINT(-11,':00*n.')
      RETURN
      END
      SUBROUTINE OUTHEX(BYTE,CHECK0)
      INTEGER BYTE,CHECK0
      INTEGER HEX(17)
      DATA HEX/176,177,178,179,180,181,182,183,184,185,193,194,195,196,1
     *97,198,0/
      CALL PUTCH(HEX(RT(RS(BYTE,4),4)+1),-11)
      CALL PUTCH(HEX(RT(BYTE,4)+1),-11)
      CHECK0=RT(CHECK0+BYTE,8)
      RETURN
      END
      SUBROUTINE GETBY0(B,FD)
      INTEGER B,FD
      INTEGER JUNK
      INTEGER MAPFD
      CALL PRWF$$(:1,MAPFD(FD),LOC(B),1,INTL(0),JUNK,JUNK)
      RETURN
      END
      SUBROUTINE GETWO0(W,FD)
      INTEGER W,FD
      INTEGER HI,LO
      CALL GETBY0(LO,FD)
      CALL GETBY0(HI,FD)
      W=OR(LS(HI,8),LO)
      RETURN
      END
C ---- Long Name Map ----
C default                        defau0
C address                        addre0
C getbyte                        getby0
C relocate                       reloc0
C getword                        getwo0
C checksum                       check0
C unnecessary                    unnec0